datafusion/expr-common/src/interval_arithmetic.rs

```diff
@@ -645,34 +651,78 @@ impl Interval {
         let upper = min_of_bounds(&self.upper, &rhs.upper);
 
         // New lower and upper bounds must always construct a valid interval.
-        assert!(
+        debug_assert!(
             (lower.is_null() || upper.is_null() || (lower <= upper)),
             "The intersection of two intervals can not be an invalid interval"
         );
 
         Ok(Some(Self { lower, upper }))
     }
 
-    /// Decide if this interval certainly contains, possibly contains, or can't
-    /// contain a [`ScalarValue`] (`other`) by returning `[true, true]`,
-    /// `[false, true]` or `[false, false]` respectively.
+    /// Compute the union of this interval with the given interval.
     ///
     /// NOTE: This function only works with intervals of the same data type.
     ///       Attempting to compare intervals of different data types will lead
     ///       to an error.
-    pub fn contains_value<T: Borrow<ScalarValue>>(&self, other: T) -> Result<bool> {
+    pub fn union<T: Borrow<Self>>(&self, other: T) -> Result<Self> {
```

Comment:
I added a `union` function in #14523 too and tried to preserve known bounds as much as possible. 

https://github.com/apache/datafusion/blob/aaf116b7a96fac7a1a4ffe902cfbb85d6be67ac2/datafusion/expr-common/src/interval_arithmetic.rs#L625 

To me the semantics of null bounds are a bit fuzzy and could be interpreted as an "indefinite" or "unknown" bound. I lean towards the latter and I think your solution is more on the former side.

Reply:
A null bound represents unboundedness for its endpoint (and is treated that way throughout the library). Seems like we should improve the documentation for more clarity to make sure it is obvious.